src/nspawn/nspawn.c

```diff
@@ -2218,6 +2285,29 @@ static int setup_machine_id(const char *directory) {
         return 0;
 }
 
+static int recursive_chown(const char *directory, uid_t shift, uid_t range) {
+        int r;
+
+        assert(directory);
+
+        if (arg_userns_mode == USER_NAMESPACE_NO || !arg_userns_chown)
+                return 0;
+
+        r = path_patch_uid(directory, arg_uid_shift, arg_uid_range);
+        if (r == -EOPNOTSUPP)
+                return log_error_errno(r, "Automatic UID/GID adjusting is only supported for UID/GID ranges starting at multiples of 2^16 with a range of 2^16.");
+        if (r == -EBADE)
+                return log_error_errno(r, "Upper 16 bits of root directory UID and GID do not match.");
+        if (r < 0)
+                return log_error_errno(r, "Failed to adjust UID/GID shift of OS tree: %m");
```

Comment:
hmm, strange, it's really weird that the code opens the dir, and that works, and then it suddenly gives up with EROFS...

Any chance you can debug this maybe with the "test-patch-uid" test tool you find in the build tree? Just pass three arguments: the os tree to patch, the start uid and the nr of uids. THe latter must be 65536 actually, and the former must be a multiple of 65536.

```
./test-patch-uid /some/path 65536 65536
```

Note that this call becomes a NOP of the root dir already has the right owner. Hence you should use a different multiple of 65536 each time for the uid start. in bash:

```
./test-patch-uid /some/path $(($RANDOM*65536))
```